moshi/src/main/java/com/squareup/moshi/Moshi.kt

```diff
@@ -0,0 +1,388 @@
+    }
+    var lookupChain = lookupChainThreadLocal.get()
+    if (lookupChain == null) {
+      lookupChain = LookupChain()
+      lookupChainThreadLocal.set(lookupChain)
+    }
+    var success = false
+    val adapterFromCall = lookupChain.push<T>(cleanedType, fieldName, cacheKey)
+    try {
+      if (adapterFromCall != null) return adapterFromCall
+
+      // Ask each factory to create the JSON adapter.
+      for (i in factories.indices) {
+        @Suppress("UNCHECKED_CAST")
+        val result = factories[i].create(cleanedType, annotations, this) as JsonAdapter<T>? ?: continue
+
+        // Success! Notify the LookupChain so it is cached and can be used by re-entrant calls.
+        lookupChain.adapterFound(result)
+        success = true
+        return result
+      }
+      throw IllegalArgumentException("No JsonAdapter for ${typeAnnotatedWithAnnotations(type, annotations)}")
+    } catch (e: IllegalArgumentException) {
+      throw lookupChain.exceptionWithLookupStack(e)
+    } finally {
+      lookupChain.pop(success)
+    }
+  }
+
+  @CheckReturnValue // Factories are required to return only matching JsonAdapters.
+  public fun <T> nextAdapter(
+    skipPast: JsonAdapter.Factory,
+    type: Type,
+    annotations: Set<Annotation>
+  ): JsonAdapter<T> {
+    val cleanedType = removeSubtypeWildcard(canonicalize(type))
+    val skipPastIndex = factories.indexOf(skipPast)
+    require(skipPastIndex != -1) { "Unable to skip past unknown factory $skipPast" }
+    for (i in (skipPastIndex + 1) until factories.size) {
+      @Suppress("UNCHECKED_CAST")
+      val result = factories[i].create(cleanedType, annotations, this) as JsonAdapter<T>?
+      if (result != null) return result
+    }
+    throw IllegalArgumentException("No next JsonAdapter for ${typeAnnotatedWithAnnotations(cleanedType, annotations)}")
+  }
+
+  /** Returns a new builder containing all custom factories used by the current instance. */
+  @CheckReturnValue
+  public fun newBuilder(): Builder {
+    val result = Builder()
+    // Runs to reuse var names
+    run {
+      val limit = lastOffset
+      for (i in 0 until limit) {
+        result.add(factories[i])
+      }
+    }
+    run {
+      val limit = factories.size - BUILT_IN_FACTORIES.size
+      for (i in lastOffset until limit) {
+        result.addLast(factories[i])
+      }
+    }
+    return result
+  }
+
+  /** Returns an opaque object that's equal if the type and annotations are equal. */
+  private fun cacheKey(type: Type, annotations: Set<Annotation>): Any {
+    return if (annotations.isEmpty()) type else listOf(type, annotations)
+  }
+
+  public class Builder {
+    internal val factories: MutableList<JsonAdapter.Factory> = ArrayList()
+    internal var lastOffset = 0
+
+    public fun <T> add(type: Type, jsonAdapter: JsonAdapter<T>): Builder = apply {
+      add(
+        newAdapterFactory(
+          type,
+          jsonAdapter
+        )
+      )
```

Comment:
Or was there a reason for this formatting?
```suggestion
      add(newAdapterFactory(type, jsonAdapter))
```